src/prof_recent.c

```diff
@@ -0,0 +1,243 @@
+
+	/*
+	 * Increment ref_count to hold the tctx so that it won't be gone even
+	 * after tctx->tdata->lock is released.  This acts as a "placeholder";
+	 * the real recording of the allocation requires a lock on
+	 * prof_recent_alloc_mtx and is done in prof_recent_alloc_slow (when
+	 * tctx->tdata->lock has been released).
+	 */
+	++tctx->ref_count;
+}
+
+static void
+decrement_ref_count(tsd_t *tsd, prof_tctx_t *tctx) {
+	/*
+	 * tdata->lock and prof_recent_alloc_mtx should not be held
+	 * simultaneously.
+	 */
+	malloc_mutex_assert_not_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+
+	malloc_mutex_lock(tsd_tsdn(tsd), tctx->tdata->lock);
+	assert(tctx->ref_count > 0);
+	--tctx->ref_count;
+	if (prof_tctx_should_destroy(tsd_tsdn(tsd), tctx)) {
+		prof_tctx_destroy(tsd, tctx);
+	} else {
+		malloc_mutex_unlock(tsd_tsdn(tsd), tctx->tdata->lock);
+	}
+}
+
+bool
+prof_recent_alloc(tsd_t *tsd, const void *ptr, size_t usize,
+    prof_tctx_t *tctx) {
+	/*
+	 * tdata->lock and prof_recent_alloc_mtx should not be held
+	 * simultaneously.
+	 */
+	malloc_mutex_assert_not_owner(tsd_tsdn(tsd), tctx->tdata->lock);
+
+	prof_recent_alloc_t *node;
+	prof_tctx_t *head_tctx;
+
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+
+	if (prof_recent_alloc_max == 0) {
+		malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+		/*
+		 * The increment of ref_count in prof_recent_alloc_fast was a
+		 * mistake - fix it.
+		 */
+		decrement_ref_count(tsd, tctx);
+		return false;
+	}
+
+	if (prof_recent_alloc_max != -1 &&
+	    prof_recent_alloc_count == prof_recent_alloc_max) {
+		/* If the upper limit is reached, simply reuse the head node. */
+		assert(prof_recent_alloc_head != NULL);
+		assert(prof_recent_alloc_tail != NULL);
+		assert(prof_recent_alloc_count > 0);
+		node = prof_recent_alloc_head;
+		prof_recent_alloc_head = node->next;
+		if (prof_recent_alloc_head == NULL) {
+			prof_recent_alloc_tail = NULL;
+		}
+		--prof_recent_alloc_count;
+		head_tctx = node->tctx;
+	} else {
+		/* Otherwise allocate a new node. */
+		node = (prof_recent_alloc_t *)iallocztm(tsd_tsdn(tsd),
+		    sizeof(prof_recent_alloc_t),
+		    sz_size2index(sizeof(prof_recent_alloc_t)),
+		    false, NULL, true, arena_get(TSDN_NULL, 0, true), true);
+		if (unlikely(node == NULL)) {
+			malloc_mutex_unlock(tsd_tsdn(tsd),
+			    &prof_recent_alloc_mtx);
+			/*
+			 * The increment of ref_count in prof_recent_alloc_fast
+			 * was a mistake - fix it.
+			 */
+			decrement_ref_count(tsd, tctx);
+			return true;
+		}
+		head_tctx = NULL;
+	}
+
+	/* Ready to add the new node. */
+	assert(prof_recent_alloc_max == -1 || (prof_recent_alloc_max > 0 &&
+	    prof_recent_alloc_count < prof_recent_alloc_max));
+	assert(node != NULL);
+
+	/* Fill in node content. */
+	nstime_t alloc_time = prof_alloc_time_get(tsd_tsdn(tsd), ptr, NULL);
+	node->alloc_time = nstime_ns(&alloc_time);
+	node->usize = usize;
+	node->tctx = tctx;
+
+	/* Append to the list */
```

Comment:
The list operation can be extracted out. Does it help to have a doubly linked list?